File: server/kafka-backend/src/core/ports/in/transfer.use-case.ts
// ============================================================
// Transfer Use Case Port (Inbound Port — Driving Side)
// Hexagonal Architecture — Ports (Core Domain)
// ============================================================

import {
  Transfer,
  TransferInitiatedEvent,
} from '../../domain/transfer.entity';

/**
 * Request DTO for initiating a transfer via REST.
 */
export interface InitiateTransferRequest {
  fromIban: string;
  toIban: string;
  amount: number;
  currency: string;
  reason: string;
  beneficiaryName: string;
  senderName: string;
}

/**
 * Response DTO returned to the HTTP client.
 */
export interface InitiateTransferResponse {
  trackingId: string;
  status: Transfer['status'];
  message: string;
}

/**
 * Primary port (driving) that the application layer implements
 * and the inbound adapter (REST controller) calls.
 */
export abstract class TransferUseCase {
  /**
   * Validates the transfer request and publishes a `transfer.initiated`
   * event to Kafka. Returns immediately with a tracking ID.
   */
  abstract initiate(
    request: InitiateTransferRequest,
  ): Promise<InitiateTransferResponse>;

  /**
   * Returns the event payload that was published.
   * Useful for idempotency checks or debugging.
   */
  abstract getLastPublishedEvent(): TransferInitiatedEvent | null;
}

File: server/java-backend/src/main/java/com/intbank/core/port/in/TransferUseCase.java
package com.intbank.core.port.in;

import com.intbank.core.domain.event.TransferInitiatedEvent;
import com.intbank.core.domain.vo.TransferStatus;

public interface TransferUseCase
{

    InitiateTransferResponse initiate(InitiateTransferRequest request);

    TransferInitiatedEvent getLastPublishedEvent();

    record InitiateTransferRequest(
        String fromIban,
        String toIban,
        double amount,
        String currency,
        String reason,
        String beneficiaryName,
        String senderName
    )
    {
    }

    record InitiateTransferResponse(
        String trackingId,
        TransferStatus status,
        String message
    )
    {
    }
}